Lib/_pylong.py

```diff
@@ -210,6 +210,117 @@ def inner(a, b):
     w5pow = compute_powers(len(s), 5, DIGLIM)
     return inner(0, len(s))
 
+# Asymptotically faster version, using the C decimal module. See
+# comments at the end of the file. This uses decimal arithmetic to
+# convert from base 10 to base 256. The latter is just a string of
+# bytes, which CPython can convert very efficiently to a Python int.
+
+import math
+_LOG_10_BASE_256 = math.log(10, 256)
+del  math
+
+def _dec_str_to_int_inner(s):
+    BYTELIM = 512
+    D = decimal.Decimal
+    result = bytearray()
+
+    def inner(n, w):
+        #assert n < D256 ** w # required, but too expensive to check
+        if w <= BYTELIM:
+            # XXX Stefan Pochmann discovered that, for 1024-bit ints,
+            # `int(Decimal)` took 2.5x longer than `int(str(Decimal))`.
+            # So simplify this code to the former if/when that gets
+            # repaired.
+            result.extend(int(str(n)).to_bytes(w)) # big-endian default
+            return
+        w2 = w >> 1
+        if 0:
+            # This is maximally clear, but "too slow". `decimal`
+            # division is asymptotically fast, but we have no way to
+            # tell it to reuse the high-precision reciprocal it computes
+            # for pow256[w2], so it has to recompute it over & over &
+            # over again :-(
+            hi, lo = divmod(n, pow256[w2][0])
+        else:
+            p256, recip = pow256[w2]
+            # The integer part will have a number of digits about equal
+            # to the difference between the log10s of `n` and `pow256`
+            # (which, since these are integers, is roughly approximated
+            # by `.adjusted()`). That's the working precision we need,
+            # but add some guard digits to protect against the "about"
+            # and "roughly" uncertainties.
+            ctx.prec = max(n.adjusted() - p256.adjusted(), 0) + 8
+            hi = +n * +recip # unary `+` chops back to ctx.prec digits
+            ctx.prec = decimal.MAX_PREC
+            hi = hi.to_integral_value() # lose the fractional digits
+            lo = n - hi * p256
+            # Because we've been uniformly rounding down, `hi` is a
+            # lower bound on the correct quotient.
+            assert lo >= 0
+            # Adjust quotient up if needed. It usually isn't. In random
+            # testing, the loop body entered about one in 100 thousand
+            # cases. I never saw it need more than one iteration.
+            count = 0
+            while lo >= p256:
+                count += 1
+                # If the assert fails, chances are decent we're sooooo
+                # far off it may seem to run forever otherwise - the
+                # error analysis was fatally flawed in this case.
+                assert count < 10, (count, w, str(n))
+                lo -= p256
+                hi += 1
```

Comment:
Can this loop be replaced with `divmod()`?
```py
            if lo >= p256:
                hi2, lo = divmod(lo, p256)
                hi += hi2
```
or few steps of subtraction followed by a `divmod()` if it was not enough?

Reply:
It could be, but I don't want to. If the error analysis is wrong, we'll never find out about it then, but the code is nonetheless broken (perhaps not manifesting as wrong results, but in mysteriously inflated runtime at times).

This is the same, to me, as the earlier `assert lo >= 0`. _Analysis_ says it must be true - but I guard against mistaken analysis there too. All my "assert"s are of that nature. I do have less confidence in this one than in `lo >= 0`, though.

Reply:
But I increasingly suspect I'm wrong. We're ignoring 3/4ths of the full "multiplication tree", and accounting for ever-more details of that now suggests the ignored part _can_ have significant effects on the integer part, and the larger the input the more "guard digits" may be needed (so that we build up more of the full multiplication tree) to move the possible errors back into the fractional part.

But finding crisp bounds on how bad it can get has eluded me, and changing the number of guard digits used depending on input size is unattractive (8 works fine - no more than 1 correction - for every input I tried so far, which have ranged up to 2.5 billion digits). So I expect to take your suggestion after all. Thanks!

BTW, I doubt more testing will find "a bad case". Even a "mere" 2.5 billion digit string takes well over an hour to convert. "Bad cases" would have a high percentage of large digits in both the input string and the reciprocal, ideally all 9s (but that's impossible for actual reciprocals). That makes ignoring the low-order digits in the full multiplication tree most damaging.